### PR TITLE
[Feat] Gradient sync is turned off during gradient accumulation.

### PR DESCRIPTION
**Pull request by EquationWalker**
Gradient sync should be turned off during gradient accumulation.

[FSDPModule.set_requires_gradient_sync](https://docs.pytorch.org/docs/stable/distributed.fsdp.fully_shard.html#torch.distributed.fsdp.FSDPModule.set_requires_gradient_sync)
[DDP.set_requires_gradient_sync](https://github.com/pytorch/pytorch/blob/v2.8.0/torch/distributed/_composable/replicate.py#L163)

### CODE REVIEW

**Review by tianyu-l — CHANGES_REQUESTED** [review]
IIUC, for FSDP, there are two ways of doing this -- doing gradient sync requires extra communication but reduce the memory overhead. Usually the gradient sync should be well-hidden, and doing it by default makes the code simpler. Hence the current choice.

See more discussions here https://github.com/pytorch/torchtitan/issues/292#issuecomment-2339101594

**Comment by EquationWalker** [comment]
> IIUC, for FSDP, there are two ways of doing this -- doing gradient sync requires extra communication but reduce the memory overhead. Usually the gradient sync should be well-hidden, and doing it by default makes the code simpler. Hence the current choice.
> 
> See more discussions here [#292 (comment)](https://github.com/pytorch/torchtitan/issues/292#issuecomment-2339101594)

Thank you very much for the informative answer! I get it.
